### PR TITLE
[repo] Make InternalsVisibleTo consistent across stable & unstable builds

### LINKED ISSUE #5477: Unintentional breaking changes? `MethodAccessException` when attempting to migrate from v1.7.0 to v1.8.0-beta.1
# Bug Report

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):

- OpenTelemetry 1.8.0-beta.1
- OpenTelemetry.Exporter.OpenTelemetryProtocol 1.7.0
- OpenTelemetry.Extensions.Hosting 1.7.0

Runtime version (e.g. `net462`, `net48`, `netcoreapp3.1`, `net6.0` etc. You can
find this information from the `*.csproj` file):

`net472`

## Symptom

**What is the expected behavior?**

Since we are not talking about major version bumps on these packages, I'd expect v1.7.0 dependencies to be fully compatible with 1.8.0 ones, but apparently 1.8.0 introduced breaking changes by removing/changing public methods that are being used by the other dependencies.

**What is the actual behavior?**

`MissingMethodException` is thrown when running the application while bumping `OpenTelemetry` NuGet from v1.7.0 to v1.8.0-beta.1 while using v1.7.0 of `OpenTelemetry.Extensions.Hosting` and `OpenTelemetry.Exporter.OpenTelemetryProtocol`

> System.MethodAccessException: 'Attempt by method 'Microsoft.Extensions.DependencyInjection.OpenTelemetryServicesExtensions.AddOpenTelemetry(Microsoft.Extensions.DependencyInjection.IServiceCollection)' to access method 'OpenTelemetry.Internal.Guard.ThrowIfNull(System.Object, System.String)' failed.'

And after upgrading `OpenTelemetry.Extensions.Hosting`:
> System.MethodAccessException: 'Attempt by method 'OpenTelemetry.Metrics.OtlpMetricExporterExtensions.AddOtlpExporter(OpenTelemetry.Metrics.MeterProviderBuilder, System.String, System.Action`1<OpenTelemetry.Exporter.OtlpExporterOptions>)' to access method 'OpenTelemetry.Internal.Guard.ThrowIfNull(System.Object, System.String)' failed.'

## Reproduce

Create a self-contained project using the template of your choice, apply the
minimum required code to result in the issue you're observing.

https://github.com/julealgon/OtelBreakingChangesRepro

## Additional Context

I wanted to bump `OpenTelemetry` to v1.8.0 to leverage the new envvar-based sampler configuration without affecting anything else in our solution.

I've now updated the other 2 offending packages to 1.8.0 as well to workaround this temporarily.

Yes, I understand 1.8.0 is still a prerelease package, but figured I'd warn the team early for when the final release is published.

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #5477

## Changes

* Make `InternalsVisibleTo` consistent across stable & unstable builds.

## Details

In unstable builds experimental APIs are marked `public` and in unstable builds they are marked `internal`. To see the things which become `internal` projects which depend on experimental APIs then need `InternalsVisibleTo` to access them. This changes how shared things like `Guard` get accessed (either through internals or linked). This leads to issues like #5477 when mixing pre-release and stable builds.

For example this case:

* OpenTelemetry.Api 1.8.0-beta.1
* OpenTelemetry 1.8.0-beta.1
* OpenTelemetry.Extensions.Hosting 1.7.0

Stable `OpenTelemetry.Extensions.Hosting` expects to access `Guard` via `OpenTelemetry.Api` internals but unstable `OpenTelemetry.Api` doesn't expose internals to `OpenTelemetry.Extensions.Hosting`.

In order to make it easier to mix packages and to reduce confusion this PR makes the `InternalsVisibleTo` consistent across build types.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)

### CODE REVIEW

File: build/Common.prod.props [hunk 1]
```diff
@@ -72,10 +72,6 @@
       <DefineConstants>$(DefineConstants);EXPOSE_EXPERIMENTAL_FEATURES</DefineConstants>
     </PropertyGroup>
 
-    <ItemGroup Condition="'$(ExposeExperimentalFeatures)' != 'true'">
```
Comment (vishweshbankwar) [inline build/Common.prod.props:75]: @CodeBlanch - Why was this needed earlier?

Comment (CodeBlanch) [reply build/Common.prod.props:75]: If you look at one of the project files on the "before" side of the diff they have things like this:

```xml
    <!-- Note: When '$(ExposeExperimentalFeatures)' == 'false' these links are
    NOT required because this project sees API + SDK internals -->
    <Compile Include="$(RepoRoot)\src\Shared\Guard.cs" Link="Includes\Guard.cs" RequiresExposedExperimentalFeatures="true" />
```

Before you could conditionally link things into projects based on whether or not experimental apis are being exposed `public` or `internal`. This PR is making `InternalsVisibleTo` a consistent thing across builds so the conditional linking is no longer needed.
